baseplate.go

```diff
@@ -0,0 +1,245 @@
+package baseplate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/reddit/baseplate.go/batcherror"
+	"github.com/reddit/baseplate.go/edgecontext"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/runtimebp"
+	"github.com/reddit/baseplate.go/secrets"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// Config is a general purpose config for assembling a Baseplate server
+type Config struct {
+	// Addr is the local address to run your server on.
+	//
+	// It should be in the format "${IP}:${Port}", "localhost:${Port}",
+	// or simply ":${Port}".
+	Addr string
+
+	// Timeout is the socket connection timeout for Servers that
+	// support that.
+	Timeout time.Duration
+
+	// StopTimeout is the timeout for the Stop command for the service.
+	//
+	// If this is not set, then no timeout will be set on the Stop command.
+	StopTimeout time.Duration `yaml:"stopTimeout"`
+
+	Log     log.Config
+	Metrics metricsbp.Config
+	Secrets secrets.Config
+	Sentry  log.SentryConfig
+	Tracing tracing.Config
+}
+
+// Baseplate is the general purpose object that you build a Server on.
+type Baseplate interface {
+	io.Closer
+
+	Config() Config
+	EdgeContextImpl() *edgecontext.Impl
+	Secrets() *secrets.Store
+}
+
+// Server is the primary interface for baseplate servers.
+type Server interface {
+	// Baseplate returns the Baseplate object the server is built on.
+	Baseplate() Baseplate
+
+	// Serve should start the Server on the Addr given by the Config and only
+	// return once the Server has stopped.
+	//
+	// It is recommended that you use baseplate.Serve() rather than calling Serve
+	// directly as baseplate.Serve will manage starting your service as well as
+	// shutting it down gracefully.
+	Serve() error
+
+	// Stop should stop the server gracefully and only return after the server has
+	// finished shutting down.
+	//
+	// It is recommended that you use baseplate.Serve() rather than calling Stop
+	// directly as baseplate.Serve will manage starting your service as well as
+	// shutting it down gracefully in response to a shutdown signal.
+	Stop() error
```

Comment:
Shouldn't `Close()` take care of what `Stop` does or are these separate steps we want to differentiate?

Reply:
I didn't really want to tie the `Baseplate` to the `Server` in that way, `Close` closes out all of the non-server stuff that `Baseplate` starts up while this is for stopping the `Server`.

Reply:
Why not? :sweat_smile:

Reply:
Part of it is around separation of concerns, the `Baseplate` is what you built a `Server` on, not the manager of the `Server`.  

It also means that you either need to build the Server and Baseplate together so you can register the Server with the Baseplate closers (which means you have to have a whole `ServerFactory` system that you pass in to New and write/test more code for each Server type) _or_ each server type has to know about this dependency and register their closer with the Baseplate they get passed in.

Finally though, I just don't know that it really makes sense?  The current flow seems the most intuitive to me where you `Close` the Baseplate _after_ the `Server` is shut down, but they aren't actually directly tied together in any way, it just happens when you `defer bp.Close()` because your `main` method won't return until the server is stopped.